### PR TITLE
[QOL-1552] add upload MIME type checking

### PR DESCRIPTION
**Pull request by ThrawnCA**
- Ensure that the claimed format, file extension, and file contents match up
- Add option for whitelisting allowed MIME types

### CODE REVIEW

File: ckanext/qgov/common/plugin.py [hunk 1]
```diff
@@ -636,3 +637,15 @@ def get_validators(self):
             'valid_url': valid_url,
             'valid_resource_url': valid_resource_url
         }
+
+    # IResourceController
+
+    def before_create(self, context, data_dict):
```
Comment (duttonw) [inline ckanext/qgov/common/plugin.py:643]: should we look at breaking this into its own plugin, this plugin.py is getting very large on implements details.

File: ckanext/qgov/common/test_mime_type_validation.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# encoding: utf-8
+
+'''Tests for the ckanext.qgov extension MIME type validation.
```
Comment (duttonw) [inline ckanext/qgov/common/test_mime_type_validation.py:3]: can't this new functionality be its own tiny plugin?

File: ckanext/qgov/common/plugin.py (hunk 1 above)
Comment (ThrawnCA) [reply ckanext/qgov/common/plugin.py:643]: It was actually supposed to be doing something like this already, but our monkey-patch stopped working when we installed ckanext-s3filestore. I suppose I could remove the obsolete code.

File: ckanext/qgov/common/test_mime_type_validation.py (hunk 2 above)
Comment (ThrawnCA) [reply ckanext/qgov/common/test_mime_type_validation.py:3]: I'm actually hoping to integrate a variation on this into core at some point (maybe just the comparison and canonicalisation of MIME type, not the whitelist).

**Review by duttonw — APPROVED** [review]

**Review by duttonw — APPROVED** [review]

**Comment by duttonw** [comment]
[Document122.pdf](https://github.com/qld-gov-au/ckan-ex-qgov/files/5799745/Document122.pdf)

demo file

**Pull request merged**